editor/src/components/editor/actions/actions.spec.browser2.tsx

```diff
@@ -1171,6 +1176,105 @@ describe('actions', () => {
         )
       })
     })
+
+    describe('end-to-end copy paste', () => {
+      const mockClipBoard: { data: any } = { data: null }
+      var pasteDone: ReturnType<typeof defer> = defer()
+      var sandbox = Sinon.createSandbox()
+
+      beforeEach(() => {
+        pasteDone = defer()
+
+        const parseClipboardDataStub = sandbox.stub(Clipboard, 'parseClipboardData')
+        parseClipboardDataStub.callsFake(async (c) => {
+          pasteDone.resolve()
+          return {
+            files: [],
+            utopiaData: extractUtopiaDataFromHtml(mockClipBoard.data),
+          }
+        })
+
+        const setClipboardDataStub = sandbox.stub(Clipboard, 'setClipboardData')
+        setClipboardDataStub.callsFake(async (c) => {
+          mockClipBoard.data = encodeUtopiaDataToHtml(c!.data)
+        })
+      })
+
+      afterEach(() => {
+        sandbox.restore()
+        mockClipBoard.data = null
+      })
```

Comment:
if this works out, you could move the beforeEach and afterEach and the variables to a separate function (in a helper file) so multiple tests can share it

Reply:
extracted it in [491c2ab](https://github.com/concrete-utopia/utopia/pull/3614/commits/491c2ab567b0ef0cde8ea8b3ce4c9602ddcc67a7)